File: ReverseAnArray.c
#include <stdio.h>

void reverseArray(int n, int arr[]) {
    // Đảo ngược mảng bằng cách hoán đổi phần tử đối xứng
    for (int i = 0; i < n / 2; i++) {
        int temp = arr[i];
        arr[i] = arr[n - i - 1];
        arr[n - i - 1] = temp;
    }
    // Hiển thị mảng sau khi đảo ngược
    printf("Mảng sau khi đảo ngược: {");
    for (int i = 0; i < n; i++) {
        if (i == n - 1)
            printf("%d", arr[i]); // Không in dấu phẩy sau phần tử cuối
        else
            printf("%d, ", arr[i]); // In dấu phẩy sau các phần tử khác
    }
    printf("}\n");

}
int main() {
    int n;
    printf("Nhập số lượng phần tử của mảng: ");
    scanf("%d", &n);
    if (n <= 0) {
        printf("Kích thước mảng phải lớn hơn 0.\n");
        return 1; // Thoát chương trình với mã lỗi
    }
    int arr[n];

    // Yêu cầu nhập các phần tử của mảng
    printf("Nhập các phần tử của mảng:\n");
    for (int i = 0; i < n; i++) {
        printf("Phần tử %d: ", i+1);
        scanf("%d", &arr[i]);
    }

    // Gọi hàm đảo ngược mảng
    reverseArray(n, arr);

    return 0; // Thoát chương trình thành công
}

File: ReverseAnArray.py
def reverseArray(arr):
    # Sử dụng slicing để đảo ngược mảng
    # arr[::-1] trả về một mảng mới với các phần tử đảo ngược
    return arr[::-1]

# Hàm đảo ngược mảng thủ công (Cách sử dụng vòng lặp)
def reverseArrayManual(arr):
    # Sử dụng hai con trỏ: 'left' bắt đầu từ đầu mảng,
    # 'right' bắt đầu từ cuối mảng
    left, right = 0, len(arr) - 1
    while left < right:
        # Hoán đổi hai phần tử ở vị trí 'left' và 'right'
        arr[left], arr[right] = arr[right], arr[left]
        # Di chuyển con trỏ vào giữa
        left += 1
        right -= 1
    # Trả về mảng đã được đảo ngược
    return arr

# Yêu cầu người dùng nhập mảng
# Tách chuỗi thành danh sách số nguyên
input_str = input("Nhập các phần tử của mảng, cách nhau bởi dấu cách: ")
arr = list(map(int, input_str.split()))

# Gọi hàm reverseArray và reverseArrayManual, đồng thời in kết quả
print("Mảng sau khi đảo ngược (slicing):", reverseArray(arr))
print("Mảng sau khi đảo ngược thủ công:", reverseArrayManual(arr))
